ArduCopter/sensors.cpp

```diff
@@ -207,6 +207,62 @@ void Copter::read_battery(void)
     }
 }
 
+// Energy based battery failsafe calculations
+//  - Returns the legacy FS_BATT_MAH value if the FS_BATT_RESERVE parameter is 0 or RTL is not possible
+//  - Calculates watts to RTL based on estimated resting voltage and hover wattage values, adds reserve watts, returns calculated MAH
+float Copter::rtl_mah_calc(void)
+{
+    // If FS_BATT_RESERVE is disabled, or HOVER_WATT is 0, or no home position available, or EKF is bad, return the default MAH
+    if (!(g2.fs_batt_reserve) || !(g2.hover_watt) || ap.home_state == HOME_UNSET || !position_ok()) { 
+        return g.fs_batt_mah;
+    };
+
+    // Power required during various phases calculated as a percentage of hover power
+    uint16_t fs_rtl_watts_hover = g2.hover_watt;                                // hover watts
+    uint16_t fs_rtl_watts_climb = fs_rtl_watts_hover * g2.fs_batt_mult_clm;     // climb watts
+    uint16_t fs_rtl_watts_desc = fs_rtl_watts_hover * g2.fs_batt_mult_dsc;      // descent watts
+    uint16_t fs_rtl_watts_cruise = fs_rtl_watts_hover * g2.fs_batt_mult_crs;    // cruise watts
+
+    // Altitude calculations
+    int32_t target_alt_cm = MAX(current_loc.alt + g.rtl_climb_min, g.rtl_altitude); // Altitude copter will RTL at
```

Comment:
Must we take into account RTL_CONE_SLOPE also?

Reply:
I'm not sure if that's something I can easily extract information from the use in the calculations. I wasn't planning to try.  If you're that close to home to begin with, the difference would (I think) be negligible anyway.

Reply:
Yes, agree. 
The other minor thing (may be you're already refactored that) that it is looks strange to have a function that does calculate RTL wattage inside sensors.cpp file.
So if there will be a future changes in RTL behavior, developer can forget to reflect them in sensors.cpp.
May be extract that calculations to control_rtl.cpp?